File: code/dobot_quad_sdk-main/low_level/python/e7_voice_pub.py
#!/usr/bin/env python3
"""
Voice Command Publisher

Supports two modes:
1. file mode: publish local file paths
2. streaming mode: capture audio from microphone in real-time and publish
"""

import sys
import subprocess
import time
import threading
import queue
import dds_middleware_python as dds


class AudioCaptureThread(threading.Thread):
    """Background thread for continuous low-latency audio capture"""

    def __init__(self, chunk_duration_ms=100):
        super().__init__(daemon=True)
        self.chunk_duration_ms = chunk_duration_ms
        self.audio_queue = queue.Queue(maxsize=2)  # Small buffer to avoid lag
        self.running = True

    def run(self):
        """Continuously capture audio chunks from microphone"""
        duration_sec = self.chunk_duration_ms / 1000.0

        try:
            # Start arecord process that will run continuously
            process = subprocess.Popen(
                ["arecord", "-q", "-t", "raw", "-f", "S16_LE", "-c", "1", "-r", "24000"],
                stdout=subprocess.PIPE,
                bufsize=0  # Unbuffered for lowest latency
            )

            # Calculate bytes per chunk (24kHz, 16bit, mono = 2 bytes per sample)
            bytes_per_chunk = int(24000 * duration_sec * 2)

            while self.running:
                chunk = process.stdout.read(bytes_per_chunk)
                if not chunk:
                    break

                # Drop old chunks if queue is full to maintain real-time performance
                try:
                    self.audio_queue.put_nowait(bytearray(chunk))
                except queue.Full:
                    # Remove oldest chunk and add new one
                    try:
                        self.audio_queue.get_nowait()
                    except queue.Empty:
                        pass
                    try:
                        self.audio_queue.put_nowait(bytearray(chunk))
                    except queue.Full:
                        pass

        except Exception as e:
            print(f"Audio capture thread error: {e}")
        finally:
            try:
                process.terminate()
            except:
                pass

    def get_audio(self):
        """Get next available audio chunk (non-blocking)"""
        try:
            return self.audio_queue.get_nowait()
        except queue.Empty:
            return None

    def stop(self):
        """Stop the capture thread"""
        self.running = False


# def capture_audio_chunk():
#     """
#     Capture ~100ms of PCM audio from microphone using arecord (24kHz, mono, 16bit)
#     """
#     try:
#         # Capture 100ms of audio (shorter duration = less latency)
#         result = subprocess.run(
#             ["arecord", "-q", "-t", "raw", "-f", "S16_LE", "-c", "1", "-r", "24000", "-d", "0.1"],
#             capture_output=True,
#             timeout=1
#         )
#         return bytearray(result.stdout)
#     except Exception as e:
#         print(f"Microphone capture failed: {e}")
#         return bytearray()


def main():
    mode = sys.argv[1] if len(sys.argv) > 1 else "file"

    # Create DDS middleware instance
    middleware = dds.PyDDSMiddleware(0)

    # Configure QoS
    qos_config = {
        "reliability": "reliable",
        "history_kind": "keep_last",
        "history_depth": 5,
        "durability": "volatile"
    }

    # Create VoiceCmd publisher
    middleware.createVoiceCmdWriter("rt/voice/cmd", qos_config)

    print(f"Mode: {mode}")
    print("QoS: RELIABLE, KEEP_LAST(5), VOLATILE")

    if mode == "file":
        # file_path = "/root/test1.wav"
        file_path = "/root/test2.flac"
        # file_path = "/root/test3.mp3"

        print("File mode: publish local file paths cyclically")
        voice_cmd = dds.VoiceCmd()
        voice_cmd.type("file")
        voice_cmd.path(file_path)
        voice_cmd.data([])
        # sleep 1 second for dds discovery
        time.sleep(1)
        middleware.publishVoiceCmd(voice_cmd)

        print("Published VoiceCmd (file)")
        print(f"  Path: {voice_cmd.path()}")
        print(f"  Data size: 0 bytes")
        print("---------------------------")


    elif mode == "streaming":
        print("Streaming mode: capture and publish from microphone (low-latency)")

        # Start background audio capture thread
        capture_thread = AudioCaptureThread(chunk_duration_ms=100)
        capture_thread.start()

        try:
            while True:
                # Get audio from capture queue (non-blocking)
                audio = capture_thread.get_audio()

                if audio is None:
                    # No audio available, sleep briefly and try again
                    time.sleep(0.01)  # 10ms instead of 200ms
                    continue

                voice_cmd = dds.VoiceCmd()
                voice_cmd.type("streaming")
                voice_cmd.path("")
                voice_cmd.data(list(audio))

                middleware.publishVoiceCmd(voice_cmd)

                print("Published VoiceCmd (streaming)")
                print(f"  Data size: {len(voice_cmd.data())} bytes")
                print("---------------------------")
                # No sleep - publish immediately when audio is available

        except KeyboardInterrupt:
            print("Stopping streaming...")
        finally:
            capture_thread.stop()

    else:
        print("Unknown mode, use 'file' or 'streaming'")


if __name__ == "__main__":
    main()

File: code/dobot_quad_sdk-main/low_level/cpp/e7_voice_pub.cc
#include "dds_middleware.hpp"
#include "voice_cmd.hpp"
#include <chrono>
#include <cstdio>
#include <iostream>
#include <string>
#include <thread>
#include <vector>
#include <queue>
#include <mutex>
#include <atomic>

using namespace dds_middleware;
using namespace dobotmh4::msg::dds_;

// Thread-safe queue for audio buffers
class AudioQueue
{
public:
    AudioQueue(size_t max_size = 2)
        : max_size_(max_size)
    {
    }

    void push(std::vector<uint8_t> data)
    {
        std::lock_guard<std::mutex> lock(mutex_);
        if (queue_.size() >= max_size_) {
            queue_.pop(); // Drop oldest if full
        }
        queue_.push(std::move(data));
    }

    bool try_pop(std::vector<uint8_t>& data)
    {
        std::lock_guard<std::mutex> lock(mutex_);
        if (queue_.empty()) {
            return false;
        }
        data = std::move(queue_.front());
        queue_.pop();
        return true;
    }

private:
    std::queue<std::vector<uint8_t>> queue_;
    std::mutex mutex_;
    size_t max_size_;
};

// Background thread for continuous low-latency audio capture
class AudioCaptureThread
{
public:
    AudioCaptureThread(int chunk_duration_ms = 100)
        : chunk_duration_ms_(chunk_duration_ms)
        , running_(true)
    {
    }

    void run()
    {
        int duration_sec_factor = chunk_duration_ms_; // e.g., 100 = 0.1 seconds

        FILE* pipe = popen("arecord -q -t raw -f S16_LE -c1 -r24000", "r");
        if (!pipe) {
            std::cerr << "arecord not found or failed to start" << std::endl;
            return;
        }

        // Calculate bytes per chunk (24kHz, 16bit, mono = 2 bytes per sample)
        // For 100ms: 24000 * 0.1 * 2 = 4800 bytes
        int bytes_per_chunk = (24000 * chunk_duration_ms_ / 1000) * 2;

        std::vector<uint8_t> buffer(bytes_per_chunk);
        size_t read_bytes = 0;

        while (running_) {
            read_bytes = fread(buffer.data(), 1, bytes_per_chunk, pipe);
            if (read_bytes > 0) {
                std::vector<uint8_t> chunk(buffer.begin(), buffer.begin() + read_bytes);
                audio_queue_.push(std::move(chunk));
            } else if (feof(pipe)) {
                break;
            }
        }

        pclose(pipe);
    }

    bool get_audio(std::vector<uint8_t>& data) { return audio_queue_.try_pop(data); }

    void stop() { running_ = false; }

private:
    AudioQueue audio_queue_;
    int chunk_duration_ms_;
    std::atomic<bool> running_;
};

// Capture audio chunk from microphone using arecord (24kHz, mono, 16bit PCM)
static std::vector<uint8_t> capture_audio_chunk()
{
    std::vector<uint8_t> data;
    FILE* pipe = popen("arecord -q -t raw -f S16_LE -c1 -r24000 -d0.1", "r");
    if (!pipe) {
        std::cerr << "arecord not found or failed to start, unable to perform streaming" << std::endl;
        return data;
    }

    std::vector<uint8_t> buffer(4096);
    size_t read_bytes = 0;
    while ((read_bytes = fread(buffer.data(), 1, buffer.size(), pipe)) > 0) {
        data.insert(data.end(), buffer.begin(), buffer.begin() + static_cast<long>(read_bytes));
    }
    pclose(pipe);
    return data;
}

int main(int argc, char** argv)
{
    std::string mode = (argc > 1) ? argv[1] : "file"; // "file" or "streaming"

    std::shared_ptr<DDSMiddleware> middleware = std::make_shared<DDSMiddleware>(0);

    QoSProfile qos;
    qos.reliability = ReliabilityPolicy::RELIABLE;
    qos.history = HistoryPolicy::KEEP_LAST;
    qos.history_depth = 5; // Align with dds_publisher.py
    qos.durability = DurabilityPolicy::VOLATILE;

    auto publisher = middleware->create_publisher<VoiceCmd_>("rt/voice/cmd", qos);

    std::cout << "Mode: " << mode << std::endl;
    std::cout << "QoS: RELIABLE, KEEP_LAST(5), VOLATILE" << std::endl;

    if (mode == "file") {
        // std::string file_path = "/root/test1.wav";
        std::string file_path = "/root/test2.flac";
        // std::string file_path = "/root/test3.mp3";

        std::cout << "File mode: publish local file paths cyclically" << std::endl;
        VoiceCmd_ voice_cmd;
        voice_cmd.type("file");
        voice_cmd.path(file_path);
        voice_cmd.data().clear();
        sleep(1);
        publisher->publish(voice_cmd);

        std::cout << "Published VoiceCmd (file)" << std::endl;
        std::cout << "  Path: " << voice_cmd.path() << std::endl;
        std::cout << "  Data size: 0 bytes" << std::endl;
        std::cout << "---------------------------" << std::endl;

        std::this_thread::sleep_for(std::chrono::seconds(1));
    }

    if (mode == "streaming") {
        std::cout << "Streaming mode: capture and publish from microphone (low-latency)" << std::endl;

        // Start background audio capture thread
        AudioCaptureThread capture_thread(100); // 100ms chunks
        std::thread audio_thread(&AudioCaptureThread::run, &capture_thread);

        std::vector<uint8_t> audio;
        while (true) {
            // Get audio from capture queue (non-blocking)
            if (capture_thread.get_audio(audio)) {
                VoiceCmd_ voice_cmd;
                voice_cmd.type("streaming");
                voice_cmd.path("");
                voice_cmd.data(audio);

                publisher->publish(voice_cmd);

                std::cout << "Published VoiceCmd (streaming)" << std::endl;
                std::cout << "  Data size: " << voice_cmd.data().size() << " bytes" << std::endl;
                std::cout << "---------------------------" << std::endl;
                // No sleep - publish immediately when audio is available
            } else {
                // No audio available, brief sleep before retry
                std::this_thread::sleep_for(std::chrono::milliseconds(10)); // 10ms instead of 200ms
            }
        }

        capture_thread.stop();
        audio_thread.join();
    }

    std::cout << "Unknown mode, use 'file' or 'streaming'" << std::endl;
    return 0;
}